File: tools/e2e.jl
# tools/e2e.jl
#
# End-to-end verification harness (specification.md §6, docs/testing.md).
# Runs every scenario in specification.md §6.1 as a subprocess and checks
# every item in specification.md §6.2 that applies to each:
#   - smoke-fast:      always run, no golden-sample check (fast sanity pass). Same
#                      -cc/-t/input as smoke-verified below by design (spec §6.1's
#                      own table pairs them that way) — this scenario is the quick
#                      pass/fail signal without paying for the golden-sample diff.
#   - invalid-input:   always run, no golden sample needed (rejected pre-computation).
#   - smoke-verified:  always run; compares against the committed golden sample
#                      test/80mm-test-ball-cc20t4-golden-sample.step via
#                      golden_sample_volume_diff.
#   - dense-lattice:   only runs if test/test-cylinder-cc10t1.5-golden-sample.step
#                      is present on disk. The `-golden-sample` suffix (matching
#                      80mm-test-ball-cc20t4-golden-sample.step) keeps a
#                      reviewed baseline clearly distinct from the ordinary
#                      run output a generation at these params produces.
#                      Originally specified
#                      at -cc 5 -t 1, an attempt to produce that golden sample
#                      ran for hours and was manually terminated mid-run (not a
#                      crash — see docs/algorithm.md §11.2's investigation): an
#                      auto-tuned tile size well past the fuse-time knee left
#                      assembly with 11k+ unfused solids, and the (now-fixed)
#                      unconditional sub-threshold cleanup rule was deleting
#                      connected junction material one solid at a time. Root
#                      cause diagnosed and fixed (docs/algorithm.md §7.1,
#                      §6.3, §8, §11.2). The scenario's params were changed to
#                      -cc 10 -t 1.5 (a less dense lattice than -cc 5 -t 1,
#                      capable of finishing within a reasonable time) with a
#                      60-minute runtime budget. This scenario stays gated on
#                      the golden sample's presence rather than being auto-run
#                      here, since committing a golden sample is a decision
#                      for whoever reviews the regenerated run's output, not
#                      something this harness should do unattended.
#
# Usage:
#   julia --project=. tools/e2e.jl

using Test

const ROOT = normpath(joinpath(@__DIR__, ".."))
include(joinpath(ROOT, "src", "latticegen2.jl"))
using .latticegen2
include(joinpath(@__DIR__, "verify_geometry.jl"))

"""
    sub_threshold_removal_ratio(output_step_path) -> Float64

Parse `<output>.log`'s mandatory end-of-run summary line ("Solids written: N
(removed as sub-threshold: M)", always printed regardless of `-v` —
`log_summary`/`log_always`, docs/algorithm.md §9) and return `M / (N + M)`.

Regression check for the root cause investigated in docs/algorithm.md
§11.2: an unconverged upstream fuse used to make the old, unconditional
"volume < threshold -> delete" cleanup rule delete large numbers of
*connected* junction solids, not genuine floating debris. A healthy run's
sub-threshold-removal ratio should be a small fraction of total solids —
this check alone would have caught the `test-cylinder-cc5t1` run (3740+
removals against a handful of kept solids) long before its multi-hour tail.
"""
function sub_threshold_removal_ratio(output_step_path::AbstractString)
    log_path = output_step_path[1:end-5] * ".log"   # ".step" is 5 chars, per cli.jl's resolve_output_paths
    isfile(log_path) || return NaN
    text = read(log_path, String)
    m = match(r"Solids written: (\d+) \(removed as sub-threshold: (\d+)\)", text)
    m === nothing && return NaN
    written, removed = parse(Int, m[1]), parse(Int, m[2])
    total = written + removed
    return total == 0 ? 0.0 : removed / total
end

# A top-level `@testset` that finishes with any failures prints its summary
# and then *throws* — uncaught, that aborts the whole script (a `LoadError`),
# so a failure in an earlier scenario would silently prevent every later
# scenario from running at all. Each scenario's `@testset` is therefore run
# inside its own try/catch below (mirroring the `global`-inside-try/catch
# pattern src/main.jl already uses for its own top-level state), so all four
# scenarios always run and are reported regardless of one another's outcome;
# `all_ok` tracks whether the run as a whole should exit nonzero.
all_ok = true

const INPUT = joinpath(ROOT, "test", "80mm-test-ball.step")
const OUTDIR = mktempdir()
const OUTPUT = joinpath(OUTDIR, "smoke-fast.step")
const BUDGET_SECONDS = 10 * 60.0   # spec §6.1: "generation < 10 minutes"
const CC = 20.0
const T = 4.0

println("=== latticegen2 e2e: smoke-fast ===")
println("Input:  ", INPUT)
println("Output: ", OUTPUT)

# specification.md §6.1's literal smoke-fast row is `-i ... -cc 20 -t 4 -bg`.
# It doesn't specify --workers/--tile-cells (--cores/--ram, added to the CLI
# after this row was written, would work too — specification.md §2/§3:
# optimization parameters are mandatory, either auto-tuned from hardware or
# given explicitly); this harness pins explicit values instead so the
# scenario is reproducible independent of the host's actual core count.
cmd = `julia --project=$ROOT $(joinpath(ROOT, "src", "main.jl")) -i $INPUT -cc $CC -t $T -o $OUTPUT --workers 4 --tile-cells 6`
println("Command: ", cmd)

t0 = time()
proc = run(pipeline(cmd; stdout=stdout, stderr=stderr); wait=false)
wait(proc)
elapsed = time() - t0

try
    @testset "smoke-fast e2e" begin
        @testset "process exits 0 with expected console output" begin
            @test proc.exitcode == 0
        end

        @testset "runtime under performance budget" begin
            println("Elapsed: $(round(elapsed, digits=1))s (budget $(BUDGET_SECONDS)s)")
            @test elapsed < BUDGET_SECONDS
        end

        @testset "STEP file written and non-empty" begin
            @test isfile(OUTPUT)
            isfile(OUTPUT) && @test filesize(OUTPUT) > 0
        end

        @testset "sub-threshold removals stay a small fraction of solids (docs/algorithm.md §8, §11.2)" begin
            ratio = sub_threshold_removal_ratio(OUTPUT)
            println("Sub-threshold removal ratio: $(round(ratio*100, digits=2))%")
            @test !isnan(ratio)
            isnan(ratio) || @test ratio < 0.01
        end

        if isfile(OUTPUT) && filesize(OUTPUT) > 0
            mesh_for_checks = with_gmsh() do
                new_model("e2e-mesh")
                import_shapes(OUTPUT)
                tessellate_surface(LatticeParams(CC, T))
            end

            @testset "round-trip read succeeds" begin
                vols = Int32[]
                with_gmsh() do
                    new_model("e2e-verify")
                    vs = import_shapes(OUTPUT)
                    append!(vols, Int32[t for (_, t) in vs])
                end
                @test !isempty(vols)
            end

            @testset "geometry validity (manifold, no self-intersections)" begin
                mesh = mesh_for_checks
                manifold_ok, bad_edges = manifold_check(mesh)
                @test manifold_ok
                manifold_ok || println("Non-manifold edges: ", length(bad_edges), " (showing up to 5): ",
                                        first(bad_edges, min(5, length(bad_edges))))

                no_self_int, bad_pairs = self_intersection_check(mesh)
                @test no_self_int
                no_self_int || println("Self-intersecting triangle pairs: ", length(bad_pairs))
            end

            @testset "output bounding box matches input within tolerance" begin
                in_lo, in_hi = with_gmsh() do
                    new_model("e2e-input-bbox")
                    import_shapes(INPUT)
                    bounding_box()
                end
                out_lo, out_hi = with_gmsh() do
                    new_model("e2e-output-bbox")
                    import_shapes(OUTPUT)
                    bounding_box()
                end
                tol = CC + T
                @test out_lo.x >= in_lo.x - tol && out_hi.x <= in_hi.x + tol
                @test out_lo.y >= in_lo.y - tol && out_hi.y <= in_hi.y + tol
                @test out_lo.z >= in_lo.z - tol && out_hi.z <= in_hi.z + tol
            end
        end
    end
catch e
    global all_ok = false
    println("smoke-fast e2e: testset reported failures (see summary above); continuing to next scenario.")
end

rm(OUTDIR; recursive=true, force=true)

println()
println("=== latticegen2 e2e: invalid-input ===")

# specification.md §6.1 `invalid-input` scenario: strut size > cell size.
# Mirrors test/test_cli.jl's "t vs cell edge cross-constraint" boundary
# (cc=5 -> a=cc/√2≈3.5355mm, t=4mm >= a) but exercised as a full subprocess
# run rather than a direct parse_args() call, so this also proves the
# process-level exit code / no-file-written behavior in src/main.jl, not
# just the validation logic in src/cli.jl.
const INVALID_OUTDIR = mktempdir()
const INVALID_OUTPUT = joinpath(INVALID_OUTDIR, "invalid-input.step")
const INVALID_LOG = joinpath(INVALID_OUTDIR, "invalid-input.log")

invalid_cmd = `julia --project=$ROOT $(joinpath(ROOT, "src", "main.jl")) -i $INPUT -cc 5 -t 4 -o $INVALID_OUTPUT --workers 1 --tile-cells 4`
println("Command: ", invalid_cmd)

invalid_stderr = IOBuffer()
invalid_proc = run(pipeline(invalid_cmd; stdout=devnull, stderr=invalid_stderr); wait=false)
wait(invalid_proc)
invalid_stderr_text = String(take!(invalid_stderr))

try
    @testset "invalid-input e2e" begin
        @testset "process exits nonzero with human-readable reason" begin
            @test invalid_proc.exitcode != 0
            @test occursin("ERROR", invalid_stderr_text)
        end

        @testset "no output STEP file written" begin
            @test !isfile(INVALID_OUTPUT)
        end

        @testset "no log file written (rejected before log is opened)" begin
            @test !isfile(INVALID_LOG)
        end
    end
catch e
    global all_ok = false
    println("invalid-input e2e: testset reported failures (see summary above); continuing to next scenario.")
end

rm(INVALID_OUTDIR; recursive=true, force=true)

println()
println("=== latticegen2 e2e: smoke-verified ===")

# specification.md §6.1 `smoke-verified` scenario. Uses -cc 20 -t 4 (not the
# originally-written -cc 10 -t 2 — that mismatched the golden sample's own
# filename, test/80mm-test-ball-cc20t4.step; -cc 20 -t 4 is what that file was
# actually generated with, so the scenario's params were corrected to match
# the golden sample rather than the other way around, per user decision).
const VERIFIED_GOLDEN = joinpath(ROOT, "test", "80mm-test-ball-cc20t4-golden-sample.step")
const VERIFIED_OUTDIR = mktempdir()
const VERIFIED_OUTPUT = joinpath(VERIFIED_OUTDIR, "smoke-verified.step")
const VERIFIED_BUDGET_SECONDS = 20 * 60.0
const VERIFIED_CC = 20.0
const VERIFIED_T = 4.0

verified_cmd = `julia --project=$ROOT $(joinpath(ROOT, "src", "main.jl")) -i $INPUT -cc $VERIFIED_CC -t $VERIFIED_T -o $VERIFIED_OUTPUT --workers 4 --tile-cells 6`
println("Command: ", verified_cmd)

verified_t0 = time()
verified_proc = run(pipeline(verified_cmd; stdout=stdout, stderr=stderr); wait=false)
wait(verified_proc)
verified_elapsed = time() - verified_t0

try
    @testset "smoke-verified e2e" begin
        @testset "process exits 0" begin
            @test verified_proc.exitcode == 0
        end

        @testset "runtime under performance budget" begin
            println("Elapsed: $(round(verified_elapsed, digits=1))s (budget $(VERIFIED_BUDGET_SECONDS)s)")
            @test verified_elapsed < VERIFIED_BUDGET_SECONDS
        end

        @testset "STEP file written and non-empty" begin
            @test isfile(VERIFIED_OUTPUT)
            isfile(VERIFIED_OUTPUT) && @test filesize(VERIFIED_OUTPUT) > 0
        end

        @testset "sub-threshold removals stay a small fraction of solids (docs/algorithm.md §8, §11.2)" begin
            ratio = sub_threshold_removal_ratio(VERIFIED_OUTPUT)
            println("Sub-threshold removal ratio: $(round(ratio*100, digits=2))%")
            @test !isnan(ratio)
            isnan(ratio) || @test ratio < 0.01
        end

        if isfile(VERIFIED_OUTPUT) && filesize(VERIFIED_OUTPUT) > 0
            @testset "geometry validity (manifold, no self-intersections)" begin
                mesh = with_gmsh() do
                    new_model("e2e-verified-mesh")
                    import_shapes(VERIFIED_OUTPUT)
                    tessellate_surface(LatticeParams(VERIFIED_CC, VERIFIED_T))
                end
                manifold_ok, bad_edges = manifold_check(mesh)
                @test manifold_ok
                manifold_ok || println("Non-manifold edges: ", length(bad_edges))

                no_self_int, bad_pairs = self_intersection_check(mesh)
                @test no_self_int
                no_self_int || println("Self-intersecting triangle pairs: ", length(bad_pairs))
            end

            @testset "matches golden sample (near-zero symmetric-difference volume)" begin
                @test isfile(VERIFIED_GOLDEN)
                if isfile(VERIFIED_GOLDEN)
                    diff_vol = golden_sample_volume_diff(VERIFIED_OUTPUT, VERIFIED_GOLDEN)
                    tol = VERIFIED_T^3   # negligible relative to model volume; spec §5's own floating-body floor
                    println("Golden-sample symmetric-difference volume: $(diff_vol) mm^3 (tolerance $(tol) mm^3)")
                    @test diff_vol < tol
                end
            end
        end
    end
catch e
    global all_ok = false
    println("smoke-verified e2e: testset reported failures (see summary above); continuing to next scenario.")
end

rm(VERIFIED_OUTDIR; recursive=true, force=true)

println()
println("=== latticegen2 e2e: dense-lattice ===")

# specification.md §6.1 `dense-lattice` scenario. Gated on the golden sample
# already existing on disk — not because the scenario is broken (it runs to
# completion; the earlier long-run and mesh-coverage problems are both
# root-caused and fixed, docs/algorithm.md §11.2 and §11.3) but because
# committing a golden sample is a review decision for whoever owns the test
# fixture, not something this harness should make for them.
const DENSE_INPUT = joinpath(ROOT, "test", "test-cylinder.STEP")
const DENSE_GOLDEN = joinpath(ROOT, "test", "test-cylinder-cc10t1.5-golden-sample.step")
const DENSE_CC = 10.0
const DENSE_T = 1.5
const DENSE_BUDGET_SECONDS = 60 * 60.0   # less dense than the original -cc 5 -t 1 scenario; 60-minute budget for now

if !isfile(DENSE_GOLDEN)
    println("SKIPPED: golden sample not found at $DENSE_GOLDEN.")
    println("         The scenario itself runs to completion — the earlier multi-hour run and the")
    println("         later mesh-coverage rejection are both root-caused and fixed (docs/algorithm.md")
    println("         §11.2, §11.3). Generating this file and committing it as the golden sample is a")
    println("         review decision, not automated here. See specification.md §6.1/§9.")
else
    const DENSE_OUTDIR = mktempdir()
    const DENSE_OUTPUT = joinpath(DENSE_OUTDIR, "dense-lattice.step")

    dense_cmd = `julia --project=$ROOT $(joinpath(ROOT, "src", "main.jl")) -i $DENSE_INPUT -cc $DENSE_CC -t $DENSE_T -o $DENSE_OUTPUT --cores 6 --ram 20`
    println("Command: ", dense_cmd)

    dense_t0 = time()
    dense_proc = run(pipeline(dense_cmd; stdout=stdout, stderr=stderr); wait=false)
    wait(dense_proc)
    dense_elapsed = time() - dense_t0

    try
        @testset "dense-lattice e2e" begin
            @testset "process exits 0" begin
                @test dense_proc.exitcode == 0
            end

            @testset "runtime under performance budget" begin
                println("Elapsed: $(round(dense_elapsed, digits=1))s (budget $(DENSE_BUDGET_SECONDS)s)")
                @test dense_elapsed < DENSE_BUDGET_SECONDS
            end

            @testset "STEP file written and non-empty" begin
                @test isfile(DENSE_OUTPUT)
                isfile(DENSE_OUTPUT) && @test filesize(DENSE_OUTPUT) > 0
            end

            @testset "sub-threshold removals stay a small fraction of solids (docs/algorithm.md §8, §11.2)" begin
                ratio = sub_threshold_removal_ratio(DENSE_OUTPUT)
                println("Sub-threshold removal ratio: $(round(ratio*100, digits=2))%")
                @test !isnan(ratio)
                isnan(ratio) || @test ratio < 0.01
            end

            if isfile(DENSE_OUTPUT) && filesize(DENSE_OUTPUT) > 0
                @testset "geometry validity (manifold, no self-intersections)" begin
                    mesh = with_gmsh() do
                        new_model("e2e-dense-mesh")
                        import_shapes(DENSE_OUTPUT)
                        tessellate_surface(LatticeParams(DENSE_CC, DENSE_T))
                    end
                    manifold_ok, bad_edges = manifold_check(mesh)
                    @test manifold_ok
                    manifold_ok || println("Non-manifold edges: ", length(bad_edges))

                    no_self_int, bad_pairs = self_intersection_check(mesh)
                    @test no_self_int
                    no_self_int || println("Self-intersecting triangle pairs: ", length(bad_pairs))
                end

                @testset "matches golden sample (near-zero symmetric-difference volume)" begin
                    diff_vol = golden_sample_volume_diff(DENSE_OUTPUT, DENSE_GOLDEN)
                    tol = DENSE_T^3
                    println("Golden-sample symmetric-difference volume: $(diff_vol) mm^3 (tolerance $(tol) mm^3)")
                    @test diff_vol < tol
                end
            end
        end
    catch e
        global all_ok = false
        println("dense-lattice e2e: testset reported failures (see summary above).")
    end

    rm(DENSE_OUTDIR; recursive=true, force=true)
end

println()
println("=== e2e summary ===")
println("smoke-fast:      exit=$(proc.exitcode) elapsed=$(round(elapsed, digits=1))s output=$OUTPUT")
println("invalid-input:   exit=$(invalid_proc.exitcode)")
println("smoke-verified:  exit=$(verified_proc.exitcode) elapsed=$(round(verified_elapsed, digits=1))s output=$VERIFIED_OUTPUT")
println("dense-lattice:   ", isfile(DENSE_GOLDEN) ? "ran (see above)" : "skipped (no golden sample yet)")
println()
println(all_ok ? "=== e2e: ALL SCENARIOS PASSED ===" : "=== e2e: ONE OR MORE SCENARIOS FAILED (see above) ===")

exit(all_ok ? 0 : 1)

File: tools/e2e.py
"""End-to-end verification harness (specification.md §6, docs/testing.md).

Runs every scenario in specification.md §6.1 as a subprocess and applies every
applicable check from §6.2. Each scenario is independent: one failing scenario
never prevents a later one from running, and the process exit code reflects the
run as a whole.

Usage:
    python tools/e2e.py [--only smoke-fast,dense-lattice]
"""

from __future__ import annotations

import os
import re
import subprocess
import sys
import tempfile
import time

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
sys.path.insert(0, os.path.join(ROOT, "src"))
sys.path.insert(0, HERE)

import verify_geometry as vg  # noqa: E402

from latticegen2 import occ, progress  # noqa: E402
from latticegen2.runlog import STAGES  # noqa: E402

PYTHON = os.environ.get("LATTICEGEN2_PYTHON", sys.executable)
MAIN = os.path.join(ROOT, "src", "main.py")
TESTDIR = os.path.join(ROOT, "test")
BALL = os.path.join(TESTDIR, "80mm-test-ball.step")
CYLINDER = os.path.join(TESTDIR, "test-cylinder.STEP")
BALL_GOLDEN = os.path.join(TESTDIR, "80mm-test-ball-cc20t4-golden-sample.step")
CYLINDER_GOLDEN = os.path.join(TESTDIR, "test-cylinder-cc10t1.5-golden-sample.step")


class Report:
    def __init__(self, name: str):
        self.name = name
        self.checks: list[tuple[str, bool, str]] = []

    def check(self, label: str, ok: bool, detail: str = "") -> bool:
        self.checks.append((label, bool(ok), detail))
        mark = "PASS" if ok else "FAIL"
        print(f"  [{mark}] {label}" + (f" — {detail}" if detail else ""), flush=True)
        return bool(ok)

    @property
    def ok(self) -> bool:
        return all(c[1] for c in self.checks)


def run_generator(args: list[str]):
    t0 = time.time()
    proc = subprocess.run(
        [PYTHON, MAIN, *args], capture_output=True, text=True, cwd=ROOT
    )
    return proc, time.time() - t0


def geometry_checks(rep: Report, output: str, input_path: str, cc: float, t: float) -> None:
    """The §6.2 checks that apply to any successfully generated STEP file."""
    rep.check("STEP file written and non-empty",
              os.path.isfile(output) and os.path.getsize(output) > 0,
              f"{os.path.getsize(output) / 1e6:.1f} MB" if os.path.isfile(output) else "missing")
    if not (os.path.isfile(output) and os.path.getsize(output) > 0):
        return

    valid, n_solids, invalid = vg.brepcheck(output)
    rep.check("round-trip read succeeds and yields solids", n_solids > 0, f"{n_solids} solid(s)")
    rep.check("exact B-rep validity (BRepCheck_Analyzer)", valid,
              "all valid" if valid else f"invalid solids: {invalid}")

    outside_check(rep, output, input_path, t)
    rep.check("bounding box within input + (cc+t)",
              vg.bounding_box_within(output, input_path, cc + t))

    mesh = vg.mesh_of(output, cc, t)
    manifold_ok, bad_edges = vg.manifold_check(mesh)
    rep.check("mesh is a closed manifold", manifold_ok,
              f"{len(mesh.tris)} triangles" + ("" if manifold_ok else f", {bad_edges} bad edges"))
    no_cross, pairs = vg.self_intersection_check(mesh)
    rep.check("no self-intersections", no_cross,
              "0 crossing pairs" if no_cross else f"{len(pairs)}+ crossing pairs")


def outside_check(rep: Report, output: str, input_path: str, t: float) -> None:
    """§6.2's "no generated material lies outside the input body", exactly where
    that is available and by a labelled weaker check where it is not.

    The exact figure is a boolean cut per output solid. On a large part the cut
    can mis-classify — it returned the whole solid as outside on a 43,530-face
    lattice, having genuinely run rather than declined — so `vg.material_outside`
    contradicts any non-zero remainder against a boolean-free containment check
    and reports the solid as unmeasured when the two disagree. Reporting that as
    a pass would be exactly the failure this check exists to catch, so it is
    reported as its own named check with the containment evidence attached.
    """
    result = vg.material_outside(output, input_path)
    if result["exact"]:
        rep.check("no material outside the input body",
                  abs(result["volume_mm3"]) < (t ** 3) * 1e-6,
                  f"{result['volume_mm3']:.6g} mm^3 over {result['solids']} solid(s)")
        return

    measured = sum(s["cut_mm3"] for s in result["per_solid"] if s["trusted"])
    n_ok = sum(1 for s in result["per_solid"] if s["trusted"])
    rep.check("no material outside the input body (solids the cut could measure)",
              abs(measured) < (t ** 3) * 1e-6,
              f"{measured:.6g} mm^3 over {n_ok} of {result['solids']} solid(s)")
    for item in result["unmeasured"]:
        ev = item["containment"]
        print(f"  [NOTE] solid {item['solid']} ({item['faces']} faces): the cut claimed "
              f"{item['cut_claimed_mm3']:.6g} mm^3 of its own {item['volume_mm3']:.6g} mm^3 "
              f"lies outside, which a boolean-free containment check contradicts; "
              f"falling back to that check")
        rep.check(f"solid {item['solid']} is contained in the input "
                  f"(weaker than the exact cut)",
                  bool(ev.get("contained")),
                  f"{ev['outside']} of {ev['sampled']} surface points outside at "
                  f"{vg.CONTAINMENT_SWEEP[0]:g} mm, none at "
                  f"{ev['cleared_at_mm']:g} mm (bar {vg.CONTAINMENT_TOL:g} mm)"
                  if ev.get("cleared_at_mm") is not None else
                  f"{ev['outside']} of {ev['sampled']} surface points still outside at "
                  f"{vg.CONTAINMENT_TOL:g} mm")


GOLDEN_EXACT_BUDGET_S = 1800.0
"""How long the exact boolean comparison gets before falling back.

Not a correctness knob, and deliberately generous. The exact test is a general
boolean between two complete lattices, so its cost grows far faster than the
lattice does: seconds on the ~10 k-face smoke output, **306 s** on the
~100 k-face dense one. An earlier 300 s budget missed that by six seconds and
sent a scenario that passes exactly (0 mm³) down the weaker sampled path — the
budget exists to stop a runaway, not to decide the common case. The fallback
never converts an unmeasured result into a pass.
"""


def golden_check(rep: Report, output: str, golden: str, cc: float, t: float) -> None:
    if not os.path.isfile(golden):
        print(f"  [SKIP] golden-sample comparison — {os.path.basename(golden)} not present")
        return
    tol = t ** 3
    diff = vg.golden_sample_volume_diff_bounded(output, golden, GOLDEN_EXACT_BUDGET_S)
    if diff is not None:
        rep.check("matches golden sample (exact symmetric-difference volume)", diff < tol,
                  f"{diff:.6g} mm^3 vs tolerance {tol:g} mm^3")
        return

    # The exact test did not finish. Report a quantified weaker check rather than
    # nothing — and label it, so a reader never mistakes it for the exact one.
    print(f"  [NOTE] exact golden comparison exceeded {GOLDEN_EXACT_BUDGET_S:.0f}s; "
          f"falling back to a sampled equivalence check")
    agree = vg.golden_sample_agreement(output, golden, cc, t)
    rep.check("golden sample: total volume matches", agree["volume_diff"] < tol,
              f"candidate {agree['candidate_volume']:.4f} vs golden "
              f"{agree['golden_volume']:.4f} mm^3 (diff {agree['volume_diff']:.6g})")
    rep.check("golden sample: bounding boxes match", agree["bbox_match"],
              f"worst extent difference {agree['bbox_delta']:.3g} mm "
              f"(tolerance {vg.BBOX_TOL} mm)")
    rep.check("golden sample: sampled membership agrees (weaker than the exact test)",
              agree["real_disagreements"] == 0,
              f"{agree['real_disagreements']} substantive of {agree['disagreements']} raw "
              f"disagreements over {agree['samples']} points; one sample = "
              f"{agree['resolution_mm3']:.4g} mm^3")


def scenario_smoke_fast(outdir: str) -> Report:
    rep = Report("smoke-fast")
    print(f"=== {rep.name} ===", flush=True)
    out = os.path.join(outdir, "smoke-fast.step")
    proc, elapsed = run_generator(
        ["-i", BALL, "-cc", "20", "-t", "4", "-o", out, "--cores", "4"]
    )
    rep.check("process exits 0", proc.returncode == 0, proc.stderr.strip()[-300:])
    rep.check("runtime under 10 minutes", elapsed < 600, f"{elapsed:.1f}s")
    if proc.returncode == 0:
        geometry_checks(rep, out, BALL, 20.0, 4.0)
    return rep


def scenario_smoke_verified(outdir: str) -> Report:
    rep = Report("smoke-verified")
    print(f"=== {rep.name} ===", flush=True)
    out = os.path.join(outdir, "smoke-verified.step")
    proc, elapsed = run_generator(
        ["-i", BALL, "-cc", "20", "-t", "4", "-o", out, "--cores", "4"]
    )
    rep.check("process exits 0", proc.returncode == 0, proc.stderr.strip()[-300:])
    rep.check("runtime under 20 minutes", elapsed < 1200, f"{elapsed:.1f}s")
    if proc.returncode == 0:
        geometry_checks(rep, out, BALL, 20.0, 4.0)
        golden_check(rep, out, BALL_GOLDEN, 20.0, 4.0)
    return rep


def scenario_dense_lattice(outdir: str) -> Report:
    rep = Report("dense-lattice")
    print(f"=== {rep.name} ===", flush=True)
    out = os.path.join(outdir, "dense-lattice.step")
    proc, elapsed = run_generator(
        ["-i", CYLINDER, "-cc", "10", "-t", "1.5", "-o", out, "--cores", "6"]
    )
    rep.check("process exits 0", proc.returncode == 0, proc.stderr.strip()[-300:])
    rep.check("runtime under 10 minutes", elapsed < 600, f"{elapsed:.1f}s")
    if proc.returncode == 0:
        geometry_checks(rep, out, CYLINDER, 10.0, 1.5)
        golden_check(rep, out, CYLINDER_GOLDEN, 10.0, 1.5)
    return rep


def _mask_volatile(text: str) -> str:
    """Blank the two timestamps a STEP header carries, and nothing else."""
    text = re.sub(r"'\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d'", "'<ts>'", text)
    return re.sub(r"generated=\d{4}-\d\d-\d\d \d\d:\d\d:\d\d", "generated=<ts>", text)


def scenario_progress_stream(outdir: str) -> Report:
    """The gate protecting the whole progress-event design (docs/algorithm.md §10).

    Watching a run must not change it. The event stream reaches into
    :class:`~latticegen2.runlog.RunLog` and into
    :meth:`~latticegen2.parallel.WorkerPool.run`'s dispatch loop, which is the
    loop every parallel stage goes through — so "additive" is a claim about the
    hot path of the entire pipeline, and the only way to hold it is to run the
    same scenario both ways and compare the bytes.

    Cheap enough to run on every invocation: ``smoke-fast`` is ~7 s, so this
    doubles one of the four scenarios rather than adding a new class of cost.
    """
    rep = Report("progress-stream")
    print(f"=== {rep.name} ===", flush=True)
    # Same basename in two directories, so the only thing that differs about
    # the two invocations is the flag under test.
    quiet_dir = os.path.join(outdir, "progress-quiet")
    loud_dir = os.path.join(outdir, "progress-loud")
    os.makedirs(quiet_dir, exist_ok=True)
    os.makedirs(loud_dir, exist_ok=True)
    plain = os.path.join(quiet_dir, "progress.step")
    streamed = os.path.join(loud_dir, "progress.step")
    base = ["-i", BALL, "-cc", "20", "-t", "4", "--cores", "4"]

    quiet, _ = run_generator([*base, "-o", plain])
    loud, _ = run_generator([*base, "-o", streamed, "--progress-stream"])

    rep.check("both runs exit 0", quiet.returncode == 0 and loud.returncode == 0,
              f"{quiet.returncode}/{loud.returncode} {loud.stderr.strip()[-200:]}")
    if quiet.returncode != 0 or loud.returncode != 0:
        return rep

    # 1. The output is the same file.
    left = _mask_volatile(open(plain, encoding="utf-8", errors="replace").read())
    right = _mask_volatile(open(streamed, encoding="utf-8", errors="replace").read())
    rep.check("STEP identical outside the header timestamp", left == right,
              f"{os.path.getsize(plain)} vs {os.path.getsize(streamed)} bytes")

    # 2. The log is the same log. Only the front-end's channel may differ.
    def log_body(step_path: str, own_dir: str) -> str:
        """The log with everything that legitimately varies between two runs
        blanked: clock times, elapsed times, measured memory, and the run's own
        directory. What survives is every count, every geometric figure and the
        order they appear in — which is precisely what an observer may not
        change.
        """
        with open(os.path.splitext(step_path)[0] + ".log", encoding="utf-8") as fh:
            body = fh.read()
        body = re.sub(r"^\[\d\d:\d\d:\d\d\] ", "", body, flags=re.M)
        body = re.sub(r"\d{4}-\d\d-\d\d[ T]\d\d:\d\d:\d\d", "<ts>", body)
        body = re.sub(r"\d{8}-\d{6}", "<stamp>", body)
        body = re.sub(r"\d+(\.\d+)?s\b", "<dur>", body)
        body = re.sub(r"\d+(\.\d+)? [KMGT]?B\b", "<size>", body)
        # `stitch` reports its six phase timings as bare seconds under keys
        # ending `_s`, which the duration pattern above cannot reach.
        body = re.sub(r"(_s: )[\d.]+", r"\1<dur>", body)
        return body.replace(own_dir, "<outdir>")

    rep.check(".log identical outside clock, memory and path",
              log_body(plain, quiet_dir) == log_body(streamed, loud_dir))

    # 3. And the stream it produced is actually usable — otherwise the two
    #    checks above would also pass if the flag did nothing at all.
    events, junk = [], []
    for line in loud.stdout.splitlines():
        if not line.strip():
            continue
        parsed = progress.parse_line(line)
        (events if parsed is not None else junk).append(parsed or line)

    kinds = {e["ev"] for e in events}
    rep.check("stdout is pure NDJSON", not junk, f"{len(junk)} unparsed: {junk[:2]}")
    rep.check("every stage is announced and closed",
              sum(e["ev"] == "stage_begin" for e in events) == len(STAGES)
              and sum(e["ev"] == "stage_end" for e in events) == len(STAGES),
              f"{sum(e['ev'] == 'stage_begin' for e in events)} begun")
    rep.check("the run reports itself finished",
              [e for e in events if e["ev"] == "result"][-1:] != []
              and events[-1]["ev"] == "result" and events[-1]["status"] == "ok")
    rep.check("sub-stage progress arrives", "progress" in kinds,
              f"{sum(e['ev'] == 'progress' for e in events)} tick(s)")
    rep.check("the quiet run printed no events", "@@" not in quiet.stdout
              and not any(progress.parse_line(l) for l in quiet.stdout.splitlines()))
    return rep


def scenario_invalid_input(outdir: str) -> Report:
    rep = Report("invalid-input")
    print(f"=== {rep.name} ===", flush=True)
    out = os.path.join(outdir, "invalid-input.step")
    log = os.path.join(outdir, "invalid-input.log")
    # t = 4 against cc = 5 gives a = 3.54 mm, so the strut cannot fit the cell.
    proc, _ = run_generator(["-i", BALL, "-cc", "5", "-t", "4", "-o", out])
    rep.check("exits with the parameter-error code", proc.returncode == 2, f"exit {proc.returncode}")
    rep.check("prints one human-readable reason", "FAILED:" in proc.stderr,
              proc.stderr.strip().splitlines()[0][:160] if proc.stderr.strip() else "(no stderr)")
    rep.check("no .step written", not os.path.isfile(out))
    rep.check("no .log written", not os.path.isfile(log))
    return rep


SCENARIOS = {
    "smoke-fast": scenario_smoke_fast,
    "invalid-input": scenario_invalid_input,
    "progress-stream": scenario_progress_stream,
    "smoke-verified": scenario_smoke_verified,
    "dense-lattice": scenario_dense_lattice,
}


def main(argv: list[str]) -> int:
    only = None
    if "--only" in argv:
        only = set(argv[argv.index("--only") + 1].split(","))
    occ.quiet_kernel()

    reports = []
    with tempfile.TemporaryDirectory(prefix="latticegen2-e2e-") as outdir:
        for name, fn in SCENARIOS.items():
            if only and name not in only:
                continue
            reports.append(fn(outdir))
            print(flush=True)

    print("=== e2e summary ===")
    for rep in reports:
        failed = [c[0] for c in rep.checks if not c[1]]
        status = "PASSED" if rep.ok else f"FAILED ({len(failed)}): " + "; ".join(failed)
        print(f"  {rep.name}: {status}")
    all_ok = all(r.ok for r in reports)
    print()
    print("=== e2e: ALL SCENARIOS PASSED ===" if all_ok
          else "=== e2e: ONE OR MORE SCENARIOS FAILED ===")
    return 0 if all_ok else 1


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
